File: models/Session.ts
import mongoose, { Schema, Document } from 'mongoose'
import type { SessionStatus, GamePhase } from '@/types'

const PlayerSchema = new Schema({
  socketId: String,
  name: { type: String, required: true },
  color: { type: String, required: true },
  score: { type: Number, default: 0 },
  bannedUntil: Date,
}, { _id: false })

const PlayedSongSchema = new Schema({
  cardId: Schema.Types.ObjectId,
  songIndex: Number,
}, { _id: false })

const CurrentRoundSchema = new Schema({
  cardId: Schema.Types.ObjectId,
  songIndex: { type: Number, default: 0 },
  phase: { type: String, enum: ['guess', 'reveal', 'between_songs'], default: 'guess' },
  buzzedPlayerId: String,
  exhaustedBuzzers: [String],
}, { _id: false })

export interface ISession extends Document {
  gameId: mongoose.Types.ObjectId
  status: SessionStatus
  players: mongoose.Types.DocumentArray<{
    socketId: string
    name: string
    color: string
    score: number
    bannedUntil?: Date
  }>
  playedSongs: { cardId: mongoose.Types.ObjectId; songIndex: number }[]
  currentRound?: {
    cardId: mongoose.Types.ObjectId
    songIndex: number
    phase: GamePhase
    buzzedPlayerId?: string
    exhaustedBuzzers: string[]
  }
  createdAt: Date
  updatedAt: Date
}

const SessionSchema = new Schema<ISession>(
  {
    gameId: { type: Schema.Types.ObjectId, ref: 'Game', required: true },
    status: { type: String, enum: ['lobby', 'active', 'ended'], default: 'lobby' },
    players: [PlayerSchema],
    playedSongs: [PlayedSongSchema],
    currentRound: CurrentRoundSchema,
  },
  { timestamps: true }
)

export default mongoose.models.Session || mongoose.model<ISession>('Session', SessionSchema)

File: backend/Models/Session.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace SpeedTune.Api.Models;

[BsonIgnoreExtraElements]
public class Player
{
    public string SocketId { get; set; } = "";
    public string Name { get; set; } = "";
    public string Color { get; set; } = "";
    public int Score { get; set; } = 0;
    public DateTime? BannedUntil { get; set; }
}

[BsonIgnoreExtraElements]
public class PlayedSong
{
    public string CardId { get; set; } = "";
    public int SongIndex { get; set; }
}

[BsonIgnoreExtraElements]
public class CurrentRound
{
    public string CardId { get; set; } = "";
    public int SongIndex { get; set; }
    public string Phase { get; set; } = "guess"; // "guess" | "reveal"
    public string? BuzzedPlayerId { get; set; }
    public List<string> ExhaustedBuzzers { get; set; } = new();
    public bool IsTimedOut { get; set; }
}

[BsonIgnoreExtraElements]
public class Session
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = ObjectId.GenerateNewId().ToString();

    [BsonRepresentation(BsonType.ObjectId)]
    public string GameId { get; set; } = "";
    public string Status { get; set; } = "lobby"; // "lobby" | "active" | "ended"
    public List<Player> Players { get; set; } = new();
    public List<PlayedSong> PlayedSongs { get; set; } = new();
    public CurrentRound? CurrentRound { get; set; }
    public List<string> PickOrder { get; set; } = new();
    public int CurrentPickerIndex { get; set; } = 0;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [BsonIgnore]
    public string? CurrentPickerId => PickOrder.Count > 0
        ? PickOrder[CurrentPickerIndex % PickOrder.Count]
        : null;
}
